### DISCUSSION TITLE
basic-extra-meal/deep-copy

### DISCUSSION THREAD

**Discussion by giscus** [itwanger/tobebetterjavaer-giscus#23] [Announcements]
# basic-extra-meal/deep-copy

一份通俗易懂、风趣幽默的Java学习指南，内容涵盖Java基础、Java并发编程、Java虚拟机、Java企业级开发、Java面试等核心知识点。学Java，就认准Java程序员进阶之路

https://tobebetterjavaer.com/basic-extra-meal/deep-copy.html

**Comment by xshi0001** [comment]
其实现在不推荐大家用Cloneable接口，实现比较麻烦，现在借助Apache Commons或者
springframework可以直接实现：

浅克隆：BeanUtils.cloneBean(Object obj);BeanUtils.copyProperties(S,T);
深克隆：SerializationUtils.clone(T object);
BeanUtils是利用反射原理获得所有类可见的属性和方法，然后复制到target类。
SerializationUtils.clone()就是使用我们的前面讲的序列化实现深克隆，当然你要把要克隆的类实现Serialization接口

**Comment by xshi0001** [comment]
有一个问题： 为啥浅克隆，引用对象克隆的是对象地址，而不能生成一个新的对象？
